File: web/lib/tiers.ts
/**
 * THE TIER LADDER, AS THE ENGINE ACTUALLY APPLIES IT.
 *
 * Five rungs — four deterministic and one assisted — run in order, with a
 * subject matched at an earlier rung removed from the pool before the next one
 * runs. This module is the copy for all five, and it is the same organising
 * rule as `lib/derivations.ts`: a count is never more than one interaction away
 * from what produced it. For a tier that means the requirement it enforces, the
 * defect classes that reach it, and what a zero on that row would mean.
 *
 * Sourced from ARCHITECTURE.md §5 (the ladder table, `core/matcher/tiers.py`)
 * and METRICS.md §11 (which defect class each rung resolves, measured).
 *
 * ONE HONEST SEPARATION, AND IT IS THE WHOLE REASON THIS FILE EXISTS.
 * `requires` is a property of the ENGINE: it is true of every run at every
 * scale, because it is the rule the code applies. `defects` is a MEASUREMENT
 * taken on one dataset — seed 42, deterministic-only, at three scales — and it
 * is not a property of the run being viewed. The UI labels the second as a
 * reference measurement and names its scale and seed; it never presents it as
 * something this run reported. The wire carries `tier_counts` and nothing that
 * attributes a match to a defect class, so any page that showed a live
 * decomposition would be inventing one.
 */
import { formatINR } from "./money";
import type { Metrics } from "./types";

export type TierKey = keyof Metrics["tier_counts"];

export const TIER_KEYS: readonly TierKey[] = ["T0", "T1", "T2", "T3", "LLM"];

export const DETERMINISTIC_TIERS: readonly TierKey[] = ["T0", "T1", "T2", "T3"];
export const ASSISTED_TIERS: readonly TierKey[] = ["LLM"];

/** One line, for a dense row. The full requirement is in `requires`. */
export type TierRequirement = {
  label: string;
  value: string;
  /** Why the rule is that and not something looser. */
  note?: string;
};

export type Tier = {
  key: TierKey;
  /** The rule, in one line, for the row itself. */
  rule: string;
  /** What this rung requires, field by field. */
  requires: TierRequirement[];
  /** What it takes to fall through to the next rung. */
  declines: string;
  /** Defect classes this rung resolves on the reference dataset. */
  defects: string;
  /** What a count of zero on this row would mean. Never "a bug". */
  zeroMeans: string;
};

/**
 * ±100 paise, as an integer, so the display value is formatted by the app's one
 * money formatter rather than typed out as a string with a rupee sign in it.
 */
export const TOLERANCE_PAISE = 100;
export const WINDOW_DAYS = 2;

/**
 * The tolerance, written the way every other amount in this app is written:
 * through the one money formatter, with the wire integer beside it. It is a
 * measurement in paise and it is rendered as rupees, and both halves are shown
 * because confusing the two is the single easiest mistake to make here.
 */
const TOLERANCE = `±${formatINR(TOLERANCE_PAISE)} (${TOLERANCE_PAISE} paise)`;

export const TIERS: Record<TierKey, Tier> = {
  T0: {
    key: "T0",
    rule: "Reference hit and exact arithmetic.",
    requires: [
      {
        label: "Reference",
        value: "A UTR, or a settlement id in the narration",
        note: "The only rung that needs the line to name its own settlement.",
      },
      {
        label: "Arithmetic",
        value: "Reconstructed net equals the bank credit exactly",
        note: "A settlement id proves identity. It says nothing about whether the sum closes, so the reference alone is never enough.",
      },
      { label: "Payment legs", value: "Any number" },
      { label: "Tolerance", value: "0 paise" },
      {
        label: "Date window",
        value: "None",
        note: "An explicit reference plus an exactly closing sum is not made more or less certain by a late posting.",
      },
    ],
    declines:
      "The reference resolves but the sum does not close. T0 writes the residual into the audit trail and lets the line fall through, rather than claiming it at confidence 1.00 with a net that is not the bank credit.",
    defects:
      "Every clean settlement, plus the six defect classes that damage the content of a settlement while leaving its reference intact and its sum exact: many-to-one batches, cross-period refunds, MDR plus GST rounding, the canonical copy of a duplicated leg, chargeback holds, and a settled leg whose order reference was recovered before matching.",
    zeroMeans:
      "No bank line in this run both named its settlement and closed to the paise. On a dataset with any clean settlements at all that would be surprising — this is the rung that carries the bulk of the work.",
  },

  T1: {
    key: "T1",
    rule: "Reconstruction closes exactly, from exactly one payment leg.",
    requires: [
      {
        label: "Reference",
        value: "Not required",
        note: "This rung exists for lines whose narration was stripped of the settlement id, the UTR and the entity.",
      },
      { label: "Arithmetic", value: "Reconstructed net equals the credit exactly" },
      {
        label: "Payment legs",
        value: "Exactly one",
        note: "Fee, tax, refund, chargeback, reserve and adjustment legs do not count. A settlement with one payment and two deduction legs is T1: it settles one order, and the deductions are the arithmetic rather than the batch.",
      },
      { label: "Tolerance", value: "0 paise" },
      { label: "Date window", value: `±${WINDOW_DAYS} days` },
    ],
    declines:
      "More than one unclaimed settlement satisfies the arithmetic and the window. More than one valid candidate means match nothing, and the candidate set goes into the audit trail.",
    defects:
      "Garbled narrations whose settlement has a single payment leg. Garbling is the only defect that removes the reference from a line that is still resolvable, so it is the only thing that ever reaches a reconstruction rung.",
    zeroMeans:
      "Nothing in this run lost its reference on a single-payment-leg settlement. A dataset with no garbled narrations, or one whose garbled lines all sit on multi-leg batches, reports zero here and is correct.",
  },

  T2: {
    key: "T2",
    rule: "Reconstruction closes exactly, from two or more payment legs.",
    requires: [
      { label: "Reference", value: "Not required" },
      { label: "Arithmetic", value: "Reconstructed net equals the credit exactly" },
      {
        label: "Payment legs",
        value: "Two or more",
        note: "Cardinality is the only thing separating this rung from T1. Method cannot separate them — an earlier draft defined T1 as reconstruction plus a date window, which is T2's method exactly, and T1 would have subsumed the rung the project exists for.",
      },
      { label: "Tolerance", value: "0 paise" },
      { label: "Date window", value: `±${WINDOW_DAYS} days` },
    ],
    declines:
      "The same ambiguity rule as T1. Candidacy is searched blind to cardinality and the ambiguity rule is applied to the whole candidate set afterwards — filtering the pool by leg count first would turn the T1/T2 split into a tie-breaker, and each settlement would look unique inside its own partition.",
    defects:
      "Garbled narrations whose settlement has two or more payment legs. The many-orders-to-one-credit case, with nothing in the narration to point at it.",
    zeroMeans:
      "No multi-leg batch in this run lost its reference. On a small dataset this is the ordinary result, not a rung that failed: a zero here is a measurement.",
  },

  T3: {
    key: "T3",
    rule: "The same reconstruction, accepted within tolerance.",
    requires: [
      { label: "Reference", value: "Not required" },
      {
        label: "Arithmetic",
        value: "Reconstructed net within tolerance of the credit",
        note: "The delta is recorded in the match evidence rather than discarded, so a tolerance match never looks like an exact one.",
      },
      {
        label: "Payment legs",
        value: "Any number",
        note: "Deliberately cardinality-agnostic. Restricting it to T2's cardinality would strand a single-payment-leg settlement carrying a rounding break, which is exactly the shape this rung exists for.",
      },
      {
        label: "Tolerance",
        value: TOLERANCE,
        note: "The loosest amount window in the system. The verifier is held to exactly the same one.",
      },
      { label: "Date window", value: `±${WINDOW_DAYS} days` },
    ],
    declines:
      "The residual is larger than the tolerance, or more than one candidate closes within it.",
    defects:
      "Rounding breaks. T0's reference lookup hits, T0 declines on the residual, T1 and T2 decline on a tolerance of zero, and this rung takes the line at confidence 0.80 with the delta on the record.",
    zeroMeans:
      "No line in this run needed the tolerance. Every match closed to the paise or did not close at all.",
  },

  LLM: {
    key: "LLM",
    rule: "The analyst proposed a resolution and the verifier accepted it.",
    requires: [
      {
        label: "Proposal",
        value: "A hypothesis naming one bank line and a set of legs",
        note: "The analyst only ever sees the residue the deterministic rungs declined.",
      },
      {
        label: "Verifier",
        value: "Six checks over five frozen labels, all of which must hold",
        note: "existence, exclusivity, causality, arithmetic, coherence and uniqueness, in that order, returning on the first failure. Coherence reports under the existence label because the five spellings are shared with the exception contract and frozen.",
      },
      {
        label: "Subject tie",
        value: "The hypothesis's subject must be the bank line it proposes against",
        note: "A seventh gate, owned by the accept loop rather than the verifier: the verifier cannot know a subject's type, so it never checks this. Without it a hypothesis could credit a resolution to a subject the data does not determine.",
      },
      {
        label: "Amount tolerance",
        value: TOLERANCE,
        note: "T3's tolerance, deliberately. A model proposal does not get a wider window than deterministic code would have taken.",
      },
      {
        label: "Money",
        value: "Recomputed, never taken from the model",
        note: "Every money field on an accepted match is recomputed by the matcher's own reconstruction from the legs, and the net is the bank credit. The hypothesis's own confidence is recorded in evidence and read by no check.",
      },
    ],
    declines:
      "Any one of the six checks fails, or the subject tie does not hold. The proposal, the free-text reason and the machine-readable failed check all survive onto the exception, which is where a rejection becomes visible rather than being discarded.",
    defects:
      "Whatever the deterministic rungs left that a single complete settlement can still explain. On the seeded datasets that set is empty, and the correct outcome is rejection rather than assistance.",
    zeroMeans:
      "Either no model was called on this run, or one was called and nothing it proposed survived the verifier. Those are different results and the analyst view separates them; a zero here alone does not.",
  },
};

/* ------------------------------------------------------------------ *
 * The reference decomposition
 * ------------------------------------------------------------------ */

export type DefectClass = {
  name: string;
  /** What the class tests. */
  tests: string;
  /**
   * The rungs that resolve it. Empty when nothing does — and two of the ten
   * classes are genuinely empty here, one of them deliberately.
   */
  resolvedBy: readonly TierKey[];
  /** How a class that reaches two rungs divides between them. */
  split?: string;
  /** What happens to it, when that needs more than the rung name. */
  outcome: string | null;
  /** Injected instances at 5,000 records, seed 42. */
  instances: number;
  /**
   * Bank-line subjects those instances produce, when it is not one each. Two
   * classes span two lines per instance, which is why 50 injected defects
   * become 100 exceptions.
   */
  subjects?: number;
};

/**
 * The ten defect classes, measured on seed 42 at 5,000 records, deterministic
 * only. Reproduced from METRICS.md §11.
 *
 * This is a REFERENCE MEASUREMENT of one dataset and not a property of the run
 * being viewed. Nothing on the wire attributes a match to a defect class, so
 * this table can never be recomputed live and is always labelled with the seed
 * and the scale it was taken at.
 */
export const REFERENCE_SEED = 42;
export const REFERENCE_RECORDS = 5000;

/**
 * `tier_counts` from the reference run, so the decomposition below can be
 * checked against it in the UI rather than asserted in prose. If a figure here
 * is ever edited out of agreement with DEFECT_CLASSES, the page says so.
 */
export const REFERENCE_TIER_COUNTS: Record<TierKey, number> = {
  T0: 1327,
  T1: 17,
  T2: 83,
  T3: 50,
  LLM: 0,
};

/** Matches the reference run produced, and the subjects it accounted for. */
export const REFERENCE_MATCHES = 1477;
export const REFERENCE_EXCEPTIONS = 300;
export const REFERENCE_BANK_LINES = 1677;

export const DEFECT_CLASSES: readonly DefectClass[] = [
  {
    name: "many_to_one_batch",
    tests: "The dominant shape: N orders paid as one bank credit.",
    resolvedBy: ["T0"],
    outcome: null,
    instances: 100,
  },
  {
    name: "cross_period_refund",
    tests: "A refund from the previous cycle netted into this one.",
    resolvedBy: ["T0"],
    outcome: null,
    instances: 44,
  },
  {
    name: "fee_plus_gst",
    tests: "MDR at 2.36%, then 18% GST on the MDR, with integer rounding at each step.",
    resolvedBy: ["T0"],
    outcome: null,
    instances: 50,
  },
  {
    name: "garbled_narration",
    tests: "Entity, settlement reference and UTR all stripped from the line.",
    resolvedBy: ["T1", "T2"],
    split: "T1 17, T2 83 — the split is payment-leg cardinality and nothing else.",
    outcome: null,
    instances: 100,
  },
  {
    name: "duplicate_psp_txn",
    tests: "The same economic event recorded twice.",
    resolvedBy: ["T0"],
    outcome: "The canonical copy matches; the duplicate is itemised as DUPLICATE_PSP_TXN, outside every rate's denominator.",
    instances: 100,
  },
  {
    name: "rounding_break",
    tests: "A ₹0.50 delta — the boundary between a tolerance match and an exception.",
    resolvedBy: ["T3"],
    outcome: null,
    instances: 50,
  },
  {
    name: "chargeback_hold",
    tests: "A deduction referencing no order in the register.",
    resolvedBy: ["T0"],
    outcome: null,
    instances: 50,
  },
  {
    name: "missing_order_ref",
    tests: "A settled payment leg with a null order id.",
    resolvedBy: ["T0"],
    outcome: "The order is recovered from the batch before matching, so the line reaches T0 intact.",
    instances: 50,
  },
  {
    name: "split_settlement",
    tests: "One settlement paid across two bank lines, closing neither on its own.",
    resolvedBy: [],
    outcome:
      "AMOUNT_MISMATCH. Both narrations name the settlement, so the reference lookup hits and then declines on the residual — which is exactly the other half's credit. This class is the entire distance between the match rate and 1.0, and closing it needs a new rung that can search subsets of bank lines.",
    instances: 50,
    subjects: 100,
  },
  {
    name: "ambiguous_unresolvable",
    tests: "Two settlements with an identical net on an identical date.",
    resolvedBy: [],
    outcome:
      "AMBIGUOUS_MULTI_CANDIDATE, by design. The data does not determine which settlement funded which line, so matching either would make iteration order the tie-breaker. Declining it is what the trap-capture rate measures.",
    instances: 50,
    subjects: 100,
  },
];

File: core/matcher/tiers.py
"""The deterministic matching tiers, T0 to T3 (spec 7).

Every tier below T0 reconstructs a settlement from its legs. That is
unavoidable: a bank credit is always net of fees, so no tier can match on a raw
amount. **Method therefore cannot distinguish T1 from T2** -- the distinction is
payment-leg **cardinality**, and cardinality is a label, not a filter.

Three rules are non-negotiable, in the order they cost most if they are wrong.

**1. Candidacy is cardinality-blind, and more than one candidate matches
nothing.** A subject's candidate set is every unclaimed settlement that
satisfies the arithmetic and the date window, at any payment-leg count. The
ambiguity rule is applied to that whole set. Only once a single candidate
survives does its payment-leg count decide whether the match is labelled T1 or
T2. Partitioning the pool by cardinality first turns the split into a
tie-breaker: two settlements that close the same credit but differ in
cardinality would each look unique inside their own partition, both subjects
would be matched, and every individual rule would still read as correct while
the whole point of the exercise quietly disappeared.

The mirror image is enforced too: when one settlement is the only candidate of
two different subjects, matching either would make iteration order the
tie-breaker, so neither is matched.

**2. T0 needs the arithmetic as well as the reference.** A settlement id in a
narration is evidence of *identity*, never a substitute for the sum. A
reference match whose reconstructed net does not equal the bank credit falls
through -- reaching T3 if it is within tolerance, and an exception if it is
not. Without that clause T0 claims a rounding break at confidence 1.00 and
writes a match whose `net` is not the bank line credit, breaking the invariant
in core/models.py, spec 6.3 and api/openapi.yaml.

**3. `MatchGroup.net` is the bank line credit.** On a T3 break the
reconstruction and the credit disagree by design; the residual is recorded in
evidence, never folded into the net.

The residual convention is `net - credit`, in that order, everywhere: positive
means the bank credited less than the reconstruction.
"""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass

from core.audit import AuditLog
from core.matcher.batch import BatchTotals
from core.matcher.pool import CandidatePool
from core.models import BankLine, MatchGroup

#: `settled_at` must sit within this many days of the bank line's `txn_date`
#: for a settlement to be a candidate at T1, T2 or T3 (spec 7).
WINDOW_DAYS = 2

#: T3's tolerance, in paise. A half-rupee rounding break is a match at 0.80; a
#: rupee-and-a-half break is an exception.
TOLERANCE_PAISE = 100


@dataclass(frozen=True)
class Candidate:
    settlement_id: str
    totals: BatchTotals
    delta: int  # net - credit, in that order
    payment_legs: int


class _Tier:
    """Shared machinery: candidate search, ambiguity, contest, match building.

    Subclasses supply the candidate search and, for T1 and T2, the cardinality
    *label* test that runs only after a single candidate has survived.
    """

    name: str
    confidence: float

    def _candidates(
        self, pool: CandidatePool, line: BankLine, log: AuditLog
    ) -> list[Candidate]:
        raise NotImplementedError

    def _accepts_cardinality(self, candidate: Candidate) -> bool:
        return True

    def _cardinality_rule(self) -> str:
        return "any payment-leg count"

    def _evidence(
        self, pool: CandidatePool, line: BankLine, candidate: Candidate
    ) -> list[str]:
        return []

    # --- the tier loop --------------------------------------------------------

    def match(self, pool: CandidatePool, log: AuditLog) -> list[MatchGroup]:
        proposals: dict[str, Candidate] = {}

        for line in pool.open_bank_lines():
            if line.credit is None:
                continue
            candidates = self._candidates(pool, line, log)

            if not candidates:
                self._record(
                    log,
                    line,
                    "no-candidate",
                    f"no unclaimed settlement satisfies {self.name}",
                )
                continue

            if len(candidates) > 1:
                ids = [c.settlement_id for c in candidates]
                pool.note_ambiguity(line.line_id, ids)
                # This is the line a maintainer reads to confirm that candidacy
                # is cardinality-blind, so it must show the candidates' ACTUAL
                # payment-leg counts. Appending the tier's own label rule to
                # this sentence read as though the set had been filtered by it,
                # which is the exact opposite of what happened -- and on the
                # trap the counts differ, so the claim was also false.
                shown = ", ".join(
                    f"{c.settlement_id} ({c.payment_legs} payment leg"
                    f"{'' if c.payment_legs == 1 else 's'})"
                    for c in sorted(candidates, key=lambda c: c.settlement_id)
                )
                self._record(
                    log,
                    line,
                    "ambiguous",
                    f"{len(ids)} equally valid candidates, searched at any "
                    f"payment-leg count: {shown}. None was filtered out by "
                    f"cardinality -- {self.name} labels {self._cardinality_rule()}, "
                    f"and that test runs only after a single candidate survives. "
                    f"More than one did, so this subject matches nothing",
                )
                continue

            candidate = candidates[0]
            if not self._accepts_cardinality(candidate):
                self._record(
                    log,
                    line,
                    "cardinality-declined",
                    f"unique candidate {candidate.settlement_id} has "
                    f"{candidate.payment_legs} payment leg(s); {self.name} labels "
                    f"{self._cardinality_rule()}",
                )
                continue

            proposals[line.line_id] = candidate

        return self._resolve(pool, proposals, log)

    def _resolve(
        self,
        pool: CandidatePool,
        proposals: dict[str, Candidate],
        log: AuditLog,
    ) -> list[MatchGroup]:
        """Reject any proposal whose settlement two subjects both want."""
        wanted: dict[str, list[str]] = defaultdict(list)
        for line_id, candidate in proposals.items():
            wanted[candidate.settlement_id].append(line_id)

        lines_by_id = {line.line_id: line for line in pool.bank_lines}
        matches: list[MatchGroup] = []

        for line_id, candidate in proposals.items():
            rivals = wanted[candidate.settlement_id]
            line = lines_by_id[line_id]
            if len(rivals) > 1:
                pool.note_contest(line_id, sorted(rivals))
                self._record(
                    log,
                    line,
                    "contested",
                    f"settlement {candidate.settlement_id} is the only candidate of "
                    f"{sorted(rivals)} -- matching nothing",
                )
                continue
            matches.append(self._build(pool, line, candidate, log))

        return matches

    def _build(
        self,
        pool: CandidatePool,
        line: BankLine,
        candidate: Candidate,
        log: AuditLog,
    ) -> MatchGroup:
        totals = candidate.totals
        narration = pool.narration(line)
        credit = line.credit or 0
        legs = pool.legs(candidate.settlement_id)

        evidence = [
            f"{self.name}: settlement {candidate.settlement_id} reconstructs to "
            f"{totals.net} paise from {len(legs)} legs",
            f"bank line {line.line_id} credit {credit} paise on "
            f"{line.txn_date.isoformat()}",
            f"payment legs {candidate.payment_legs} -- labelled {self.name}",
            *self._evidence(pool, line, candidate),
        ]
        if narration.entity:
            evidence.append(
                f"narration entity {narration.entity!r} recorded as evidence only; "
                f"it is not a matching criterion"
            )
        # Both of these used to be a linear filter of a whole-run list, run once
        # per match built: every order recovery, and every dangling reference
        # crossed with every leg of this settlement. The pool groups them by
        # settlement at construction instead, in the same order, so the evidence
        # block reads identically -- the audit trail is the product and its
        # wording is not an implementation detail.
        for recovery in pool.recoveries_for(candidate.settlement_id):
            evidence.append(recovery.evidence)
        for txn_id, dangling in pool.dangling_refs_for(candidate.settlement_id):
            evidence.append(
                f"{txn_id} references {dangling}, which is not in the order "
                f"register -- excluded from the order set"
            )

        match = MatchGroup(
            match_id=f"match-{line.line_id}",
            bank_line_id=line.line_id,
            settlement_id=candidate.settlement_id,
            psp_txn_ids=[t.txn_id for t in legs],
            order_ids=pool.order_ids(candidate.settlement_id),
            gross=totals.gross,
            fees=totals.fees,
            tax=totals.tax,
            refunds=totals.refunds,
            holds=totals.holds,
            # The invariant is `net == bank credit`. On a T3 break the
            # reconstruction is 50 paise away and the residual lives in
            # evidence -- it is never folded into this field.
            net=credit,
            tier=self.name,
            confidence=self.confidence,
            evidence=evidence,
        )
        pool.claim(line.line_id, candidate.settlement_id)
        log.record(
            line.line_id,
            "match",
            "deterministic",
            f"{self.name}:matched",
            "; ".join(evidence),
            self.confidence,
        )
        return match

    def _record(self, log: AuditLog, line: BankLine, rule: str, evidence: str) -> None:
        log.record(
            line.line_id, "match", "deterministic", f"{self.name}:{rule}", evidence, 0.0
        )


# --- T0 -----------------------------------------------------------------------


class _T0(_Tier):
    """Reference hit AND exact arithmetic. Confidence 1.00.

    T0 does not apply the date window: an explicit settlement reference plus an
    exactly closing sum is not made more or less certain by a late posting.
    """

    name = "T0"
    confidence = 1.00

    def _cardinality_rule(self) -> str:
        return "any payment-leg count"

    def _candidates(
        self, pool: CandidatePool, line: BankLine, log: AuditLog
    ) -> list[Candidate]:
        settlement_id = pool.referenced_settlement(line)
        if settlement_id is None:
            return []
        if pool.is_claimed(settlement_id):
            self._record(
                log,
                line,
                "reference-claimed",
                f"reference {settlement_id} hit but that settlement is already "
                f"matched to another bank line",
            )
            return []

        totals = pool.totals(settlement_id)
        delta = totals.net - (line.credit or 0)
        if delta != 0:
            self._record(
                log,
                line,
                "reference-hit-arithmetic-declined",
                f"reference {settlement_id} hit, but it reconstructs to {totals.net} "
                f"paise against credit {line.credit} -- residual delta={delta} paise "
                f"(net - credit). A settlement id proves identity, not arithmetic, "
                f"so {self.name} declines and the line falls through",
            )
            return []

        return [
            Candidate(
                settlement_id=settlement_id,
                totals=totals,
                delta=0,
                payment_legs=pool.payment_legs(settlement_id),
            )
        ]

    def _evidence(
        self, pool: CandidatePool, line: BankLine, candidate: Candidate
    ) -> list[str]:
        source = "narration" if pool.narration(line).settlement_id else "utr column"
        return [
            f"settlement reference {candidate.settlement_id} found in the {source} "
            f"and present in the PSP report",
            "reconstructed net equals the bank credit exactly (residual delta=0)",
        ]


# --- T1 / T2 / T3 -------------------------------------------------------------


class _ReconstructionTier(_Tier):
    """Shared candidate search for the tiers that rebuild from legs.

    The search is deliberately identical for T1, T2 and T3 apart from the
    tolerance. Cardinality is applied afterwards, to a set of exactly one.
    """

    tolerance: int

    def _candidates(
        self, pool: CandidatePool, line: BankLine, log: AuditLog
    ) -> list[Candidate]:
        """Every unclaimed settlement satisfying the arithmetic and the date
        window, at any payment-leg count.

        This used to walk the whole unclaimed pool for every bank line, and
        `unclaimed_settlements()` re-sorted that pool on every call -- three
        times per subject, once for each of T1, T2 and T3. It is an index
        lookup now (spec §8), and the set it produces is unchanged: the pool's
        `unclaimed_settlements_near_net` answers exactly the predicate the scan
        applied, inclusive at both tolerance edges, and returns **all** the
        settlements at a matching net rather than the first.

        That last part is the whole safety argument. The set this returns is
        what the ambiguity rule counts, so a lookup that stopped early would
        turn two candidates into one and a declined subject into a false match,
        with every rule below still reading as correct. The index is keyed on
        the net alone for the mirror-image reason: cardinality labels the
        winner and must never narrow the field.
        """
        credit = line.credit or 0
        found: list[Candidate] = []
        for settlement_id in pool.unclaimed_settlements_near_net(credit, self.tolerance):
            if not pool.within_window(settlement_id, line.txn_date, WINDOW_DAYS):
                continue
            totals = pool.totals(settlement_id)
            found.append(
                Candidate(
                    settlement_id=settlement_id,
                    totals=totals,
                    delta=totals.net - credit,
                    payment_legs=pool.payment_legs(settlement_id),
                )
            )
        return found

    def _evidence(
        self, pool: CandidatePool, line: BankLine, candidate: Candidate
    ) -> list[str]:
        settled = pool.settled_on(candidate.settlement_id)
        return [
            f"only unclaimed settlement within tolerance {self.tolerance} paise and "
            f"{WINDOW_DAYS} days, at any payment-leg count",
            f"settled_at {settled.isoformat() if settled else 'unknown'} is within "
            f"{WINDOW_DAYS} days of bank date {line.txn_date.isoformat()}",
            f"residual delta={candidate.delta} paise (net - credit)",
        ]


class _T1(_ReconstructionTier):
    """Exactly one `payment` leg; exact arithmetic; date window. Confidence 0.95.

    `fee`, `tax`, `refund`, `chargeback`, `reserve` and `adjustment` legs do not
    count toward the cardinality -- a settlement with one payment, one fee and
    one tax leg settles one order, and the deduction legs are the arithmetic,
    not the batch.
    """

    name = "T1"
    confidence = 0.95
    tolerance = 0

    def _accepts_cardinality(self, candidate: Candidate) -> bool:
        return candidate.payment_legs == 1

    def _cardinality_rule(self) -> str:
        return "exactly one payment leg"


class _T2(_ReconstructionTier):
    """Two or more `payment` legs; otherwise identical to T1. Confidence 0.99.

    The genuine many-to-one batch, and the core of the project: it is the only
    tier that solves the actual problem shape.
    """

    name = "T2"
    confidence = 0.99
    tolerance = 0

    def _accepts_cardinality(self, candidate: Candidate) -> bool:
        return candidate.payment_legs >= 2

    def _cardinality_rule(self) -> str:
        return "two or more payment legs"


class _T3(_ReconstructionTier):
    """Within tolerance at ANY payment-leg count. Confidence 0.80.

    Cardinality-agnostic on purpose: restricting T3 to T2's cardinality would
    strand a single-payment-leg settlement carrying a rounding break, which is
    exactly the shape the fixture's `rounding_break` defect has.
    """

    name = "T3"
    confidence = 0.80
    tolerance = TOLERANCE_PAISE


T0 = _T0()
T1 = _T1()
T2 = _T2()
T3 = _T3()

#: Tier order. A subject matched at an earlier tier is removed from the pool
#: before the next one runs.
TIERS: tuple[_Tier, ...] = (T0, T1, T2, T3)
